### PR TITLE
[SPARK-46850][SQL] Convert `_LEGACY_ERROR_TEMP_2102 ` to `UNSUPPORTED_DATATYPE`

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to 
- convert `_LEGACY_ERROR_TEMP_2102` to `UNSUPPORTED_DATATYPE`.
- remove some outdated comments.

### Why are the changes needed?
The changes improve the error framework.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Add new UT
- Pass GA

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
cc @MaxGekk 
When I was working on https://github.com/apache/spark/pull/44665 related work, I found that `_LEGACY_ERROR_TEMP_2102` can be converted into `UNSUPPORTED_DATATYPE`
Moreover, from the current implementation logic, some of its comment are incorrect.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala [hunk 1]
```diff
@@ -273,8 +272,7 @@ class UnivocityParser(
     case udt: UserDefinedType[_] =>
       makeConverter(name, udt.sqlType, nullable)
 
-    // We don't actually hit this exception though, we keep it for understandability
-    case _ => throw QueryExecutionErrors.unsupportedTypeError(dataType)
```
Comment (panbingkun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:277]: This exception can be triggered from the `user's perspective`, so I suggest removing it.

File: sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala [hunk 2]
```diff
@@ -159,6 +160,29 @@ class CsvFunctionsSuite extends QueryTest with SharedSparkSession {
       Row(Row(null)))
   }
 
+  test("from_csv with invalid datatype") {
```
Comment (panbingkun) [inline sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:163]: The process of triggering ``.

**Review by MaxGekk — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala [hunk 3]
```diff
@@ -829,7 +829,7 @@ private[columnar] object ColumnType {
       case map: MapType => MAP(PhysicalMapType(map.keyType, map.valueType, map.valueContainsNull))
       case struct: StructType => STRUCT(PhysicalStructType(struct.fields))
       case udt: UserDefinedType[_] => ColumnType(udt.sqlType)
-      case other => throw QueryExecutionErrors.unsupportedTypeError(other)
+      case other => throw ExecutionErrors.unsupportedDataTypeError(dataType)
```
[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala:832]:
```suggestion
      case _ => throw ExecutionErrors.unsupportedDataTypeError(dataType)
```

**Pull request closed without merging**